pubsub/google/cloud/pubsub_v1/publisher/batch/thread.py

```diff
@@ -0,0 +1,385 @@
+        self._thread = None
+
+        # Continually monitor the thread until it is time to commit the
+        # batch, or the batch is explicitly committed.
+        if autocommit and self._settings.max_latency < float('inf'):
+            self._thread = threading.Thread(target=self.monitor)
+            self._thread.start()
+
+    @property
+    def client(self):
+        """Return the client used to create this batch.
+
+        Returns:
+            ~.pubsub_v1.client.PublisherClient: A publisher client.
+        """
+        return self._client
+
+    @property
+    def messages(self):
+        """Return the messages currently in the batch.
+
+        Returns:
+            Sequence: The messages currently in the batch.
+        """
+        return self._messages
+
+    @property
+    def settings(self):
+        """Return the batch settings.
+
+        Returns:
+            ~.pubsub_v1.types.BatchSettings: The batch settings. These are
+                considered immutable once the batch has been opened.
+        """
+        return self._settings
+
+    @property
+    def size(self):
+        """Return the total size of all of the messages currently in the batch.
+
+        Returns:
+            int: The total size of all of the messages currently
+                 in the batch, in bytes.
+        """
+        return self._size
+
+    @property
+    def status(self):
+        """Return the status of this batch.
+
+        Returns:
+            str: The status of this batch. All statuses are human-readable,
+                all-lowercase strings.
+        """
+        return self._status
+
+    def commit(self):
+        """Actually publish all of the messages on the active batch.
+
+        This synchronously sets the batch status to in-flight, and then opens
+        a new thread, which handles actually sending the messages to Pub/Sub.
+
+        .. note::
+
+            This method is non-blocking. It opens a new thread, which calls
+            :meth:`_commit`, which does block.
+        """
+        # Set the status to in-flight synchronously, to ensure that
+        # this batch will necessarily not accept new messages.
+        #
+        # Yes, this is repeated in `_commit`, because that method is called
+        # directly by `monitor`.
+        self._status = 'in-flight'
+
+        # Start a new thread to actually handle the commit.
+        commit_thread = threading.Thread(target=self._commit)
+        commit_thread.start()
+
+    def _commit(self):
+        """Actually publish all of the messages on the active batch.
+
+        This moves the batch out from being the active batch to an in-flight
+        batch on the publisher, and then the batch is discarded upon
+        completion.
+
+        .. note::
+
+            This method blocks. The :meth:`commit` method is the non-blocking
+            version, which calls this one.
+        """
+        # Update the status.
+        self._status = 'in-flight'
```

Comment:
Then make them private on the enum. Magic strings are as bad as magic numbers. :)